app/src/main/java/org/kiwix/kiwixmobile/localFileTransfer/PeerGroupHandshake.kt

```diff
@@ -0,0 +1,110 @@
+/*
+ * Kiwix Android
+ * Copyright (c) 2019 Kiwix <android.kiwix.org>
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package org.kiwix.kiwixmobile.localFileTransfer
+
+import android.util.Log
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.withContext
+import org.kiwix.kiwixmobile.core.BuildConfig
+import java.io.InputStream
+import java.io.ObjectInputStream
+import java.io.ObjectOutputStream
+import java.io.OutputStream
+import java.net.InetAddress
+import java.net.InetSocketAddress
+import java.net.ServerSocket
+import java.net.Socket
+
+/**
+ * Helper class for the local file sharing module.
+ *
+ * Once two peer devices are connected through wifi direct, this task is executed to perform a
+ * handshake between the two devices. The purpose of the handshake is to allow the file sending
+ * device to obtain the IP address of the file receiving device (When the file sending device
+ * is the wifi direct group owner, it doesn't have the IP address of the peer device by default).
+ *
+ * After obtaining the IP address, the sender also shares metadata regarding the file transfer
+ * (no. of files & their names) with the receiver.
+ */
+abstract class PeerGroupHandshake(var wifiDirectManager: WifiDirectManager) {
+  private val HANDSHAKE_MESSAGE = "Request Kiwix File Sharing"
+  suspend fun handshake(): InetAddress? =
+    withContext(Dispatchers.IO) {
+      if (BuildConfig.DEBUG) {
+        Log.d(TAG, "Handshake in progress")
+      }
+      if (wifiDirectManager.isGroupFormed && wifiDirectManager.isGroupOwner && this.isActive) {
```

Comment:
this won't compile, use groupInfo not wifiDirectManager. as I said in another comment it will probably look better if you pass this in as a parameter to `handshake` instead of a constructor param